final-project/application.py

```diff
@@ -34,19 +34,94 @@ def after_request(response):
 # configure CS50 Library to use SQLite database
 db = SQL("sqlite:///final.db")
 
+# users table
+# CREATE TABLE users (id INTEGER PRIMARY KEY NOT NULL, username TEXT UNIQUE NOT NULL, email TEXT UNIQUE NOT NULL, hash TEXT NOT NULL, first TEXT, last TEXT);
+
 @app.route("/")
 @login_required
 def index():
 
     # keep user logged in
     session.get("user_id")
 
-    """Render map."""
-    if not os.environ.get("API_KEY"):
-        raise RuntimeError("API_KEY not set")
-    return render_template("index.html", key=os.environ.get("API_KEY"))
+    return apology("TODO")
+
+
+@app.route("/categories")
+@login_required
+def categories():
+
+    # keep user logged in
+    session.get("user_id")
+
+    return apology("TODO")
+
+@app.route("/product")
+# @login_required
+def product():
+
+    # keep user logged in
+    session.get("user_id")
+
+    # get name of product
+    product = "%" + request.args.get("product") + "%"
+
+    try:
+        #Get product from the database
+        product_info = db.execute("SELECT * FROM products WHERE product_name LIKE :product", product = product)
+        # Check whether products were found
+        product_check = product_info[0]
+
+    except RuntimeError:
+        # If problem with db.execute
+        return apology("Error: We'll fix this. Please try again shortly.")
+
+    except IndexError:
+        # If no products were found
+        return apology("No products named " + request.args.get("product"))
+
+    #Send data to product template
+    return render_template("product.html", product_name = product_info[0]["product_name"], image = product_info[0]["image"], link = product_info[0]["link"], description = product_info[0]["description"], brand = product_info[0]["brand"], price = usd(product_info[0]["price"]))
+    return apology("TODO")
 
-# @app.route("/articles", methods=["GET", "POST"])
+
+@app.route("/new")
+def new():
+    """Add new product"""
+
+    # remember user
+    session.get("user_id")
+
+    if request.method == "POST":
+
+        try:
+            # Add new product
+            # TODO: add request.form.get() to get parameters
+            new_product_row = db.execute("INSERT INTO products (id, category_id, product_name, link, description, image, brand, price) VALUES (NULL, 1, :product_name, :link, :description, :image, :brand, :price)", product_name = request.form.get("product_name"), link = request.form.get("link"), description = request.form.get("description"), image = request.form.get("image"), brand = request.form.get("brand"), price = request.form.get("price"))
+
+            # If a table constraint was violated
+            if new_product_row == None:
+                # render an apology
+                return apology("Something went wrong. Please try again later.")
+
+        except RuntimeError:
```

Comment:
I think that you are using way too many try catches. If there is an internal server error, you should just handle it altogether. (For 500 responses) rather than wrapping your methods in `try ... catch`es. Generally try catches are also way less performant.

Reply:
oh good to know.

Will do.

Reply:
I consolidated all the try catches to a single one per method. However, I do not think I can reduce the number of try catches beyond that. The reason there are more than in other files, is because of the db.execute function I use from the cs50 library. It is a short (and less performant) way to run the database queries. I could read about the proper way to query the database without my cs50 training wheels, but it would take more time. It is on my list of v2 work.

For the try catches which exist, I tried to optimize what I could. I read the PEP 8 Guidelines for tries and catches. They recommend only doing one thing per "try", not to include any excess code. Therefore, I updated the code to adhere to these guidelines.

Reply:
🏁 done